llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -1812,6 +1813,49 @@ static Value *simplifyOrOfICmps(ICmpInst *Op0, ICmpInst *Op1,
   return nullptr;
 }
 
+/// Test if a pair of compares with a shared operand and 2 constants has an
+/// empty set intersection, full set union, or if one compare is a superset of
+/// the other.
+static Value *simplifyAndOrOfFCmpsWithConstants(FCmpInst *Cmp0, FCmpInst *Cmp1,
+                                                bool IsAnd) {
+  // Look for this pattern: {and/or} (fcmp X, C0), (fcmp X, C1)).
+  if (Cmp0->getOperand(0) != Cmp1->getOperand(0))
+    return nullptr;
+
+  const APFloat *C0, *C1;
+  if (!match(Cmp0->getOperand(1), m_APFloat(C0)) ||
+      !match(Cmp1->getOperand(1), m_APFloat(C1)))
+    return nullptr;
+
+  auto Range0 = ConstantFPRange::makeExactFCmpRegion(Cmp0->getPredicate(), *C0);
+  auto Range1 = ConstantFPRange::makeExactFCmpRegion(Cmp1->getPredicate(), *C1);
+
+  if (!Range0 || !Range1)
+    return nullptr;
+
+  // For and-of-compares, check if the intersection is empty:
+  // (fcmp X, C0) && (fcmp X, C1) --> empty set --> false
+  if (IsAnd && (*Range0).intersectWith(*Range1).isEmptySet())
+    return getFalse(Cmp0->getType());
+
+  // For or-of-compares, check if the union is full:
+  // (fcmp X, C0) || (fcmp X, C1) --> full set --> true
+  if (!IsAnd && (*Range0).unionWith(*Range1).isFullSet())
```

Comment:
`unionWith` is not precise. It may return a superset of `Range0 U Range1`.

Reply:
Use De Morgan's Law `A || B -> !(!A && !B)`. You can use `ConstantFPRange::makeExactFCmpRegion(FCmpInst::getInvertedPredicate(Cmp0->getPredicate()))` to get the range for `!A` and `!B`. After simplification, you got `!!A -> A` or `!!B -> B`.

Reply:
> unionWith is not precise. It may return a superset of Range0 U Range1.

This same logic is also used in `simplifyAndOrOfICmpsWithConstants`, so is this specific with `fcmp`?

Reply:
Yes. ConstantFPRange doesn't support ranges like `X < -1 || X > 1` (i.e., wrapped sets in ConstantRange).